### ISSUE TITLE
[Enhancement] Add a requirement file for Python.

### ISSUE THREAD

**Issue by SeppPenner** [per1234/inolibbuglist#6]
It would be cool to have a `requirements.txt` (see e.g. https://pip.readthedocs.io/en/1.1/requirements.html) to simply install all required pip packages using `pip install -r requirements.txt`. As an example, maybe take a look at my repository: https://github.com/SeppPenner/TelegramMessageBot.

*Same as in the inoliblist project* https://github.com/per1234/inoliblist/issues/6 👍

**Comment by SeppPenner** [comment]
Well, since you're using only standard libraries, the requirements.txt file does not bring any benefits here. However, I've added it (as empty file) now.

**Comment by per1234** [comment]
Would codespell be a requirement? I'm currently installing it like this:
https://github.com/per1234/inolibbuglist/blob/7249ed29712d74686d246354b68046dd6c33019e/inolibbuglist.py#L921

If I remember correctly, I did originally try to use requirements.txt for that, but had trouble making it work.

**Comment by SeppPenner** [comment]
> Would codespell be a requirement? I'm currently installing it like this:
> [inolibbuglist/inolibbuglist.py](https://github.com/per1234/inolibbuglist/blob/7249ed29712d74686d246354b68046dd6c33019e/inolibbuglist.py#L921)
> 
> Line 921 in [7249ed2](/per1234/inolibbuglist/commit/7249ed29712d74686d246354b68046dd6c33019e)
> 
>  subprocess_run(command="pip", arguments="install codespell") 
> If I remember correctly, I did originally try to use requirements.txt for that, but had trouble making it work.

`codespell` is not in the [Python standard library](https://docs.python.org/3/library/), so yes. This could be put into the `requirements.txt` file.

**Comment by SeppPenner** [comment]
> And where is the problem now? Can we do that? I think that can't be much a problem

Check the pull request: https://github.com/per1234/inolibbuglist/issues/6#issuecomment-487503454, please. There is some error.

**Comment by TomasRoj** [comment]
What is codespell used for here?

**Comment by per1234** [comment]
>There is some error

Regarding the Travis CI build failure, there are two issues:
- [`Blank line found at start of ./requirements.txt`](https://travis-ci.org/per1234/inolibbuglist/builds/525645837#L454)/[`Blank line found at end of ./requirements.txt`](https://travis-ci.org/per1234/inolibbuglist/builds/525645837#L468): The formatting checks are failing because the empty requirements.txt file has an extra blank line. This can be fixed by simply removing the blank line.
- [`error: argument --ghtoken: expected one argument`](https://travis-ci.org/per1234/inolibbuglist/builds/525645837#L510): `$GITHUB_TOKEN` is a Travis CI secret environment variable so it is not defined during pull request builds submitted by 3rd parties (otherwise they could steal the token). This is the first time I had a contribution to these projects so I wasn't aware of the issue before now. I have a fix for this, and will push that soon.

**Comment by per1234** [comment]
>What is codespell used for here?

From the readme:
https://github.com/per1234/inolibbuglist#bug-detection
>Detect common misspellings using codespell.

When enabled, the script checks the repositories of all the Arduino Libraries in the inoliblist output list for typos and indicates whether any were found in the inolibbuglist output list.

In addition, codespell is used in the Travis CI build to check for any typos in this repository.

**Comment by TomasRoj** [comment]
@per1234 if it can be fixed by single backspace why is it still there?

**Comment by per1234** [comment]
>why is it still there?

Because I'm not sure that I'll merge this pull request. I'm not convinced of the need to add a blank requirements.txt file (though I'm open to arguments for that). Now we are talking about adding codespell to that file, which would make me more favorable to this proposal, but I'll need to see how that is implemented and test it out.

The issue with GITHUB_TOKEN is something I need to resolve regardless, so that's a higher priority for me. Even if I didn't end up merging this pull request, it has already been valuable in that it exposed a bug in the Travis CI configuration.

**Comment by SeppPenner** [comment]
> Because I'm not sure that I'll merge this pull request. I'm not convinced of the need to add a blank requirements.txt file (though I'm open to arguments for that). Now we are talking about adding codespell to that file, which would make me more favorable to this proposal, but I'll need to see how that is implemented and test it out.

I totally agree here. It does only make sense to include a requirements.txt file that uses some dependencies.

>The issue with GITHUB_TOKEN is something I need to resolve regardless, so that's a higher priority for me. Even if I didn't end up merging this pull request, it has already been valuable in that it exposed a bug in the Travis CI configuration.

I'm glad, it helped at least for that :D

**Comment by TomasRoj** [comment]
So if I am right you are trying to make requirements file to install all dependencies in one command right? And you are facing some errors with travis. So is it all? Why we need this file?

**Comment by SeppPenner** [comment]
> So if I am right you are trying to make requirements file to install all dependencies in one command right? And you are facing some errors with travis. So is it all? Why we need this file?

Exactly. We do not need it necessarily, but it's commonly used: https://pip.pypa.io/en/stable/user_guide/#requirements-files

**Comment by TomasRoj** [comment]
Ok I will be glad if I can help. Give me please some definite challenge which I can do

**Comment by TomasRoj** [comment]
Write something or something similar

**Comment by per1234** [comment]
>Write something or something similar

@TomasRoj I don't understand what you mean. Please try to be a bit more clear in your communication.

**Comment by TomasRoj** [comment]
Sorry for that. I mean if you can give me something to fix.

**Comment by per1234** [comment]
Did you look at the other open issues in this repository's issue tracker:
https://github.com/per1234/inolibbuglist/issues
You're welcome to take on one of those.

The funny thing is that the sole purpose of this project is finding things to fix, but in other repositories. It surely has served me well in that regard. I submitted ~3000 pull requests over the last year to fix the bugs found by this script, and there are plenty more still left.

**Comment by TomasRoj** [comment]
omfg. 3000 pull requests are you kidding me? I must do something similar

**Comment by SeppPenner** [comment]
@TomasRoj: @per1234 really is crazy in this kind of way :D

**Comment by TomasRoj** [comment]
Does anything similar exists for github repositories in general?

**Comment by per1234** [comment]
>I must do something similar

@TomasRoj Yes, I hope you will, but remember that quality is much more important than quantity.

I'm not sure if you read what I wrote in your forum thread about contributing about the importance of making sure that our attempts to contribute have an overall positive effect:
http://forum.arduino.cc/index.php?topic=607386.msg4121261#msg4121261
I think you would do well to review your attempts to contribute so far, consider whether each had an overall positive effect, and think about what you could do to improve that balance in future contributions.

**Comment by TomasRoj** [comment]
Yes. Back-then when I asked this question I was pretty noob (I am still but smaller 😂), I was new to this github and arduino community. From that time I learned how to commit, make pull request or open an issue. Its beautiful how open source can change whole workflow. I am pretty glad that I did some contributions which help beginners (naucse.python.cz which is biggest czech site for learning python) or similar. It gives me brand new experience, I met new people and I am pretty glad that I found someone who can help me (ex. you). Code quality is still my problem, I know that 🙌

**Comment by per1234** [comment]
>Does anything similar exists for github repositories in general?

That I know of, there is no automated scan for bugs across GitHub. A lot of the scans would be dependent on language. I think the only ones that could be adapted from this project for that purpose would be:
- Spell check using codespell
- "failure" status (indicates the CI build failed)

You can use GitHub's search feature to discover opportunities to contribute:

- https://github.com/issues?q=is%3Aopen+is%3Aissue+archived%3Afalse+label%3A%22help+wanted%22
- https://github.com/search?q=label%3Afirst-timers-only+is%3Aissue+is%3Aopen

And some other projects that do the same:
- https://github.com/yourfirstpr/yourfirstpr.github.io/issues
- https://github.com/MunGell/awesome-for-beginners

**Comment by TomasRoj** [comment]
Thanks, I idk github api but I think creating something like that in python using codespell can be pretty good idea. Do you know anyone who can help me with building this?

**Comment by per1234** [comment]
If I were you, I'd just get started by using inolibbuglist to get the list of Arduino Library repositories that fail codespell checks. Fixing all those misspelled words will keep you plenty busy for a long time to come.

After that, you can decide if you want to write your own script to expand the scan to more repositories. All the code to do that is there for you to adapt as you wish in inoliblist.py and inolibbuglist.py.

**Comment by TomasRoj** [comment]
Okay. You are experienced contributor I will do what you think. Thanks

**Comment by TomasRoj** [comment]
How can I use this? I see some commands in readme but idk how to install/search for fails

**Comment by per1234** [comment]
I'd like to refrain from taking the discussion here any farther off the topic of adding a requirements.txt file. If you need support using inoliblist or inolibbuglist, I'd prefer to do that on the Arduino forum.

**Comment by TomasRoj** [comment]
Ok posted there now
